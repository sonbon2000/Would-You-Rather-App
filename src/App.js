import React from "react";
import { Route, Routes } from "react-router-dom";
import "./App.css";
import PrivateRoutes from "./pages/PrivateRoute";
import Home from "./pages/Home/Home";
import LeaderBoard from "./pages/LeaderBoard/LeaderBoard";
import Login from "./pages/Login/Login";
import NewQuestion from "./pages/NewQuestion/NewQuestion";
import QuestionDetail from "./pages/QuestionDetail/QuestionDetail";
import { ToastContainer } from "react-toastify";
import NotFound from "./pages/NotFound/NotFound";
import { useEffect } from "react";
import { useDispatch } from "react-redux";
import { useNavigate } from "react-router-dom";
import { _getQuestions, _getUsers } from "./api/data";
import { receiveUsers } from "./store/usersSlice";
import { receiveQuestions } from "./store/questionsSlice";
function App() {
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const initializeDataApp = async () => {
    const users = await _getUsers();
    const questions = await _getQuestions();

    dispatch(receiveUsers(users));
    dispatch(receiveQuestions(questions));
  };

  useEffect(() => {
    initializeDataApp();
    localStorage.removeItem("auth");
  }, [dispatch]);

  useEffect(() => {
    const auth = localStorage.getItem("auth");
    if (!auth) {
      navigate("/login");
    }
  }, []);

  return (
    <div className="App">
      <ToastContainer />
      <Routes>
        <Route element={<PrivateRoutes />}>
          <Route path="/" element={<Home />} />
          <Route path="/add" element={<NewQuestion />} />
          <Route path="/question/:questionId" element={<QuestionDetail />} />
          <Route path="/leaderboard" element={<LeaderBoard />} />
        </Route>
        <Route path="/login" element={<Login />} />
        <Route path="*" element={<NotFound />} />
      </Routes>
    </div>
  );
}

export default App;
