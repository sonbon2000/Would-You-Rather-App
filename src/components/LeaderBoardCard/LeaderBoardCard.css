.leader-img {
    width: 110px;
}

.grid-left {
    width: 75%;
    padding: 10px 0;
}

.grid-right {
    width: 25%;
    padding: 10px 0;
}

.leaderboard-header {
    background: #f3f4f5;
    margin: 0;
    padding: 10px 0;
}

.total-score {
    background: #21ba45;
    width: 40px;
    height: 40px;
    justify-content: center;
    display: flex;
    font-weight: bold;
    align-items: center;
    font-size: 20px;
    color: white;
    border-radius: 50%;
    margin: auto;
}