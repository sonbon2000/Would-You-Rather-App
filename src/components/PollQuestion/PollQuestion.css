.poll-ques-author {
    background: #f3f4f5;
    border: 1px #d5d5d6 solid;
    padding: 10px 0;
    border-radius: 3px;
}

.poll-ques-img {
    width: 90px;
}

.poll-ques-button {
    padding: 8px 20px;
    background: #21ba45;
    color: #f1fbf3;
    border-radius: 5px;
    font-weight: 600;
}

.poll-ques-button:hover {
    cursor: pointer;
    background: #00AD28;
}

.not-submmit-poll-ques-button {
    background: #9ce0ac;
    padding: 8px 20px;
    color: #f1fbf3;
    border-radius: 5px;
    font-weight: 600;
}