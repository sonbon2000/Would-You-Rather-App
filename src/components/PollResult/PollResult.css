.poll-result-author {
    background: #f3f4f5;
    border: 1px #d5d5d6 solid;
    padding: 10px 0;
    border-radius: 3px;
}

.poll-img {
    width: 120px;
}

.poll-button {
    padding: 8px 20px;
    background: #e0e1e2;
    color: #5a5a5a;
    float: right;
    border-radius: 5px;
    font-weight: 600;
}

.poll-button:hover {
    cursor: pointer;
    background: #D5D5D6;
    color: black;
}