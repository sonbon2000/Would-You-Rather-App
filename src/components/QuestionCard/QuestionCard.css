.ques-card-header {
    background: #f3f4f5;
    border: 1px #d4d4d5 solid;
    border-radius: 3px;
    padding: 10px 0;
}

.ques-card-img { 
    width: 90px;
}

.ques-card-button {
    padding: 10px 0;
    border-radius: 5px;
    font-weight: bold;
    color: #eafdff;
}

.ques-card-button:hover {
    cursor: pointer;
}
