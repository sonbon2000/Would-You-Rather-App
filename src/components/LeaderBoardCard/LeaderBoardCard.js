import React from "react";
import PropTypes from "prop-types";
import {
  Grid,
  GridColumn,
  Label,
  Segment,
  Divider,
  SegmentGroup,
} from "semantic-ui-react";
import './LeaderBoardCard.css';

const LeaderBoardCard = ({ leaderBoardData, index }) => {
  const { name, avatarURL, answeredQuestions, createdQuestions, totalScore } =
    leaderBoardData;

  const rankingColorLabel = ["yellow", "grey", "orange"];

  return (
    <Segment>
      <Label corner="left" icon="trophy" color={rankingColorLabel[index]} />
      <Grid divided padded>
        <GridColumn width={4} verticalAlign="middle">
          <img className="leader-img" src={avatarURL} alt=""/>
        </GridColumn>

        <GridColumn width={8}>
          <h3>{name}</h3>
          <Grid>
            <div className="grid-left">Answered questions</div>
            <div className="grid-right">{answeredQuestions}</div>
          </Grid>
          <Divider />
          <Grid>
            <div className="grid-left">Created questions</div>
            <div className="grid-right">{createdQuestions}</div>
          </Grid>
        </GridColumn>

        <GridColumn width={4} textAlign="center">
          <SegmentGroup>
            <h5 className="leaderboard-header">Score</h5>
            <Segment>
              <div className="total-score">
                {totalScore}
              </div>
            </Segment>
          </SegmentGroup>
        </GridColumn>
      </Grid>
    </Segment>
  );
};

LeaderBoardCard.propTypes = {
  leaderBoardData: PropTypes.object.isRequired,
  index: PropTypes.number.isRequired,
};

export default LeaderBoardCard;
