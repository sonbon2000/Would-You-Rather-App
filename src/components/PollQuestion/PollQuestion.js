import React from "react";
import PropTypes from "prop-types";
import {
  SegmentGroup,
  Grid,
  GridColumn,
  Form,
  FormField,
  Radio,
} from "semantic-ui-react";
import "./PollQuestion.css";
import { usePollQuestion } from "./usePollQuestion";
const PollQuestion = ({ questionData, author }) => {
  const [values, handlers] = usePollQuestion({ questionData });
  return (
    <SegmentGroup>
      <h5 className="poll-ques-author">{author.name} asks:</h5>
      <Grid divided padded>
        <GridColumn width={4}>
          <img alt="" className="poll-ques-img" src={author.avatarURL} />
        </GridColumn>
        <GridColumn width={12}>
          <h5>Would You Rather?</h5>
          <Form onSubmit={handlers.handleSubmitAnswer}>
            <FormField>
              <Radio
                label={questionData.optionOne.text}
                value="optionOne"
                checked={values.selection === "optionOne"}
                onChange={handlers.handleSelectQuestion}
              />
              <br />
              <Radio
                label={questionData.optionTwo.text}
                value="optionTwo"
                checked={values.selection === "optionTwo"}
                onChange={handlers.handleSelectQuestion}
              />
            </FormField>
            <FormField>
              {values.isDisabled ? (
                <div className="not-submmit-poll-ques-button">Submit</div>
              ) : (
                <div
                  className="poll-ques-button"
                  onClick={handlers.handleSubmitAnswer}
                >
                  Submit
                </div>
              )}
            </FormField>
          </Form>
        </GridColumn>
      </Grid>
    </SegmentGroup>
  );
};

PollQuestion.propTypes = {
  questionData: PropTypes.object.isRequired,
  author: PropTypes.object.isRequired,
};

export default PollQuestion;
