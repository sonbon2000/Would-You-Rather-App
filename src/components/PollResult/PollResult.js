import React from "react";
import PropTypes from "prop-types";
import {
  SegmentGroup,
  Grid,
  GridColumn,
  Segment,
  Label,
  Icon,
  Progress,
} from "semantic-ui-react";
import { useSelector } from "react-redux";
import { useNavigate } from "react-router-dom";
import './PollResult.css';

const PollResult = ({ questionData, author }) => {

  const { userId } = useSelector((state) => state.auth);
  const navigate = useNavigate();

  const optionOneVotes = questionData.optionOne.votes.length;
  const optionTwoVotes = questionData.optionTwo.votes.length;
  const totalVotes = optionOneVotes + optionTwoVotes;

  const isVote = (option) => {
    return questionData[option].votes.includes(userId);
  };

  const vote_Color = (option) => {
    return questionData[option].votes.includes(userId) ? "green" : "grey";
  };

  const renderYourVoteLabel = () => {
    return (
      <Label color="yellow" ribbon="right">
        <Icon name="check circle outline" size="big" />
        <p style={{ float: "right" }}>Your Vote</p>
      </Label>
    );
  };

  const handleGoBack = () => {
    navigate("/");
  };

  return (
    <SegmentGroup>
      <h5 className="poll-result-author">
        {author.name} asks:
      </h5>

      <Grid divided padded>
        <GridColumn width={5} verticalAlign="middle">
          <img alt="" className="poll-img" src={author.avatarURL} />
        </GridColumn>

        <GridColumn width={11}>
          <h3>Results:</h3>

          <Segment color={vote_Color("optionOne")}>
            {isVote("optionOne") && renderYourVoteLabel()}
            <h5>{questionData.optionOne.text}</h5>
            <Progress
              percent={((optionOneVotes / totalVotes) * 100).toFixed(0)}
              progress
              color={vote_Color("optionOne")}
            />
            <p>
              {optionOneVotes} out of {totalVotes} votes
            </p>
          </Segment>

          <Segment color={vote_Color("optionTwo")}>
            {isVote("optionTwo") && renderYourVoteLabel()}
            <h5>{questionData.optionTwo.text}</h5>
            <Progress
              percent={((optionTwoVotes / totalVotes) * 100).toFixed(0)}
              progress
              color={vote_Color("optionTwo")}
            />
            <p>
              {optionTwoVotes} out of {totalVotes} votes
            </p>
          </Segment>

          <div className="poll-button" onClick={handleGoBack}>
            Go back
          </div>
        </GridColumn>
      </Grid>
    </SegmentGroup>
  );
};

PollResult.propTypes = {
  questionData: PropTypes.object.isRequired,
  author: PropTypes.object.isRequired,
};

export default PollResult;