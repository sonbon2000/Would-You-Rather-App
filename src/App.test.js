/* eslint-disable testing-library/render-result-naming-convention */
/* eslint-disable testing-library/prefer-screen-queries */
import { fireEvent, render } from "@testing-library/react";
import Login from "./pages/Login/Login";
import { Provider } from "react-redux";
import { store } from "./store";
import { MemoryRouter } from "react-router-dom";
import App from "./App";
import Home from "./pages/Home/Home";
import NewQuestion from "./pages/NewQuestion/NewQuestion";
import LeaderBoard from "./pages/LeaderBoard/LeaderBoard";
import NotFound from "./pages/NotFound/NotFound";
import QuestionDetail from "./pages/QuestionDetail/QuestionDetail";
import QuestionCard from "./components/QuestionCard/QuestionCard";
import NavBar from "./components/NavBar/Navbar";
import PollQuestion from "./components/PollQuestion/PollQuestion";
import PollResult from "./components/PollResult/PollResult";
import LeaderBoardCard from "./components/LeaderBoardCard/LeaderBoardCard";

describe("Test whole app", () => {
  const { getByText, getByRole } = render(
    <MemoryRouter>
      <Provider store={store}>
        <App />
      </Provider>
    </MemoryRouter>
  );
  test("Login page", () => {
    expect(<Login />).toMatchSnapshot();
    expect(getByText("Welcome to Would You Rather game!")).toBeInTheDocument();
    expect(getByText("Please sign in to continue")).toBeInTheDocument();
    const button = getByText("Sign In");
    fireEvent.click(button);
    expect(getByRole("alert")).toBeInTheDocument();
  });

  test("Home page", () => {
    expect(<Home />).toMatchSnapshot();
  });

  test("New Question page", () => {
    expect(<NewQuestion />).toMatchSnapshot();
  });

  test("Leaderboard page", () => {
    expect(<LeaderBoard />).toMatchSnapshot();
  });

  test("Not found page", () => {
    expect(<NotFound />).toMatchSnapshot();
  });

  test("Question Detail page", () => {
    expect(<QuestionDetail />).toMatchSnapshot();
  });

  test("Question Card comp", () => {
    expect(<QuestionCard />).toMatchSnapshot();
  });

  test("NavBar comp", () => {
    expect(<NavBar />).toMatchSnapshot();
  });

  test("PollQuestion comp", () => {
    expect(<PollQuestion />).toMatchSnapshot();
  });

  test("PollResult comp", () => {
    expect(<PollResult />).toMatchSnapshot();
  });

  test("LeaderBoardCard comp", () => {
    expect(<LeaderBoardCard />).toMatchSnapshot();
  });
});
