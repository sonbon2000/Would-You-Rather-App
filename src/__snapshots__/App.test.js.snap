// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page 1`] = `<Login />`;

exports[`Provider Login page 1`] = `<Login />`;

exports[`Test whole app Home page 1`] = `<Home />`;

exports[`Test whole app LeaderBoardCard comp 1`] = `<LeaderBoardCard />`;

exports[`Test whole app Leaderboard page 1`] = `<LeaderBoard />`;

exports[`Test whole app Login page 1`] = `<Login />`;

exports[`Test whole app NavBar comp 1`] = `<NavBar />`;

exports[`Test whole app New Question page 1`] = `<NewQuestion />`;

exports[`Test whole app New question page 1`] = `<Home />`;

exports[`Test whole app Not found page 1`] = `<NotFound />`;

exports[`Test whole app PollQuestion comp 1`] = `<PollQuestion />`;

exports[`Test whole app PollResult comp 1`] = `<PollResult />`;

exports[`Test whole app Question Card comp 1`] = `<QuestionCard />`;

exports[`Test whole app Question Detail page 1`] = `<QuestionDetail />`;
